packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,877 @@
+    public viewport: string = "";
+    private viewportChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.viewportElement = this.getViewport();
+            this.updateDimensions();
+        }
+    }
+
+    /**
+     * The size in pixels of each item along the virtualization axis
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: item-size
+     */
+    @attr({ attribute: "item-size", converter: nullableNumberConverter })
+    public itemSize: number = 50;
+    private itemSizeChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.updateDimensions();
+        }
+    }
+
+    /**
+     * Defines an area in pixels on either end of the viewport where items outside the viewport
+     * will still be rendered.
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: viewport-buffer
+     */
+    @attr({ attribute: "viewport-buffer", converter: nullableNumberConverter })
+    public viewportBuffer: number = 100;
+
+    /**
+     * Whether the list is oriented vertically or horizontally.
+     * Default is vertical
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: orientation
+     */
+    @attr
+    public orientation: Orientation = Orientation.vertical;
+    private orientationChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.updateDimensions();
+        }
+    }
+
+    /**
+     * Auto update mode defines what prompts the component to check the dimensions of elements
+     * in the DOM and reset the visible items accordingly.  Calling update() always provokes an update.
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: auto-update-mode
+     */
+    @attr({ attribute: "auto-update-mode" })
+    public autoUpdateMode: VirtualListAutoUpdateMode = "manual";
+    private autoUpdateModeChanged(
+        prevMode: VirtualListAutoUpdateMode,
+        newMode: VirtualListAutoUpdateMode
+    ): void {
+        if (this.$fastController.isConnected) {
+            this.resetAutoUpdateMode(prevMode, newMode);
+        }
+    }
+
+    /**
+     * Whether or not to recycle the html container used to display items.
+     * May help performance but containers may retain artifacts from previous use that
+     * developers will need to clear.
+     *
+     * @public
+     */
+    @attr({ attribute: "recycle", mode: "boolean" })
+    public recycle: boolean = true;
+
+    /**
+     *  The array of items to be displayed
+     *
+     * @public
+     */
+    @observable
+    public items: object[] = [];
+    private itemsChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.reset();
+        }
+    }
+
+    /**
+     * The sizemap for the items
+     * Authors need to provide a sizemap for arrays of irregular size items,
+     * when the items have a uniform size use the 'item-size' attribute instead.
+     *
+     * @public
+     */
+    @observable
+    public sizemap: SizeMap[];
```

Comment:
Do you have a practical use-case for this API? Note that I don't see any fast-component example demonstrating how to use it. My concern (as with the `itemsize` attribute) is that this is putting a heavy/unnecessary burden on the client. To me, this API only seems helpful, in a practical sense, if the various items you were rendering were grouped together in their respective sizes, which seems unlikely.

Would it better to just not support virtualization of items of various sizes, then to offer an API to satisfy what seems a pretty niche/unlikely scenario?

For instance, the WPF VirtualizingStackPanel never asks for similar information for the items its virtualizing.

Another API that considers both cases of items with the same size or uneven sizes is the [Xamarin ListView](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.listview.hasunevenrows?view=xamarin-forms), whose API more or less mirrors what I suggested about the `item-size` attribute.

Reply:
My prototype branch has a "[gallery panel](https://github.com/microsoft/fast/tree/users/scomea/virt-stack-panel/packages/web-components/fast-foundation/src/gallery-panel)" sample component that demonstrates how a size-map could be used to virtualize across nested regions.   

This would be the approach to enable virtualization of things like tree view in a way that allowed nested trees to virtualize based on the same viewport as the top level parent (Note the prototype branch is not up to date with the pr branch currently).  We have an internal React based tree-view that uses a similar approach.  

If I understand correctly the difference between this and WPF is that WPF virtualization is directly bound to the scrolling container and that container scrolls based on index of elements, not on the actual size of elements.    While the WPF approach may be more forgiving on the element size side it may not be so flexible with respect to nested virtualizing lists. 

I expect we'll want to do a "custom scroll bar behavior" variant at some point where more dynamic "render based" item sizing can work.

Reply:
Am I correct in saying that if the `fast-tree` or `fast-accordion` were to try and leverage the `virtual-list` they would not only change their templates to use the `virtual-list` (could they even use the current `slotted` approach for their content?), but would also have to manually update the `visibleItemSpans` of the `virtual-list` when expanding/collapsing sections (which means they would have to track the index of the item being expanded/collapsed)?